src/protocols/bdx/AsyncTransferFacilitator.cpp

```diff
@@ -0,0 +1,154 @@
+/*
+ *
+ *    Copyright (c) 2023 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "AsyncTransferFacilitator.h"
+
+#include <system/SystemClock.h>
+
+namespace chip {
+namespace bdx {
+
+CHIP_ERROR AsyncTransferFacilitator::OnMessageReceived(chip::Messaging::ExchangeContext * ec,
+                                                       const chip::PayloadHeader & payloadHeader,
+                                                       chip::System::PacketBufferHandle && payload)
+{
+    VerifyOrReturnError(ec == mExchange.Get(), CHIP_ERROR_INCORRECT_STATE);
+
+    ChipLogDetail(BDX, " %s: message " ChipLogFormatMessageType " protocol " ChipLogFormatProtocolId, __FUNCTION__,
+                  payloadHeader.GetMessageType(), ChipLogValueProtocolId(payloadHeader.GetProtocolID()));
+    CHIP_ERROR err =
+        mTransfer.HandleMessageReceived(payloadHeader, std::move(payload), System::SystemClock().GetMonotonicTimestamp());
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(BDX, "failed to handle message: %" CHIP_ERROR_FORMAT, err.Format());
+    }
+
+    // Almost every BDX message will follow up with a response on the exchange. Even messages that might signify the end of a
+    // transfer could necessitate a response if they are received at the wrong time.
+    // For this reason, it is left up to the application logic to call ExchangeContext::Close() when it has determined that the
+    // transfer is finished.
+    ec->WillSendMessage();
+
+    // Get the next output event and send it to the delegate.
+    TransferSession::OutputEvent outEvent;
+    do
+    {
+        mTransfer.GetNextAction(outEvent);
+
+        if (outEvent.EventType == TransferSession::OutputEventType::kTransferTimeout)
+        {
+            OnResponseTimeout(ec);
```

Comment:
I think this part is broken, as is timeout handling here in general.

In particular, we just received a message.   How can we get a timeout here?

More generally, who is tracking timeouts?  The transfer session or the exchange?  If it's the transfer session, it actually does rely on being polled off some timer or something to detect a timeout.  If it's supposed to be the exchange, why are we never setting the timeout value we have on the exchange?  Right now our exchange will never time out unless the secure session gets torn down under it.

This part really needs to get sorted out carefully, and at some point we need to add some unit tests for this stuff.

Reply:
The bdx session is tracking the timeout right now (as it has in the past). I didn't change this part. It emits the transfertimeout event so whenever I am getting next action I am checking for the transfer timeout. But I agree with you if we can set the timeout on the exchange we can handle it via the exchange itself which will be cleaner. I will test the timeout part as I planned to but haven't done that and report back. I can add the timeout to the exchange and test that out as well

Reply:
> so whenever I am getting next action I am checking for the transfer timeout. 

But what will cause you to get next action if you are just waiting for the server to get back to you?